knowledge_repo/app/auth_providers/oauth2.py

```diff
@@ -0,0 +1,165 @@
+        'scopes': 'account',
+        'user_info_endpoint': 'user',
+        'user_info_mapping': {
+            'identifier': 'username',
+            'name': 'name',
+            'photo_uri': ['links', 'avatar']
+        }
+    },
+    'github': {
+        'base_url': 'https://api.github.com/',
+        'authorization_url': 'https://github.com/login/oauth/authorize',
+        'token_url': 'https://github.com/login/oauth/access_token',
+        'auto_refresh_url': None,
+        'scopes': None,
+        'user_info_endpoint': 'user',
+        'user_info_mapping': {
+            'identifier': 'email',
+            'name': 'name',
+            'photo_uri': 'avatar_url'
+        }
+    },
+    'google': {
+        'base_url': 'https://www.googleapis.com/oauth2/v2/',
+        'authorization_url': 'https://accounts.google.com/o/oauth2/auth',
+        'token_url': 'https://accounts.google.com/o/oauth2/token',
+        'auto_refresh_url': 'https://accounts.google.com/o/oauth2/token',
+        'scopes': [
+            'https://www.googleapis.com/auth/userinfo.email',
+            'https://www.googleapis.com/auth/userinfo.profile'
+        ],
+        'user_info_endpoint': 'userinfo',
+        'user_info_mapping': {
+            'identifier': 'email',
+            'name': 'name',
+            'photo_uri': 'picture'
+        }
+    }
+}
+
+
+def _resolve_oauth_config(name, local_namespace, config, *variables):
+    presets = PRESETS.get(name, {})
+    output = []
+    for variable in variables:
+        value = local_namespace.get(variable, None)
+        if value is None:
+            value = config.get('OAUTH_{}_{}'.format(name, variable).upper(), None)
+        if value is None:
+            value = presets.get(variable, None)
+        output.append(value)
+    return output
+
+
+class OAuth2Provider(KnowledgeAuthProvider):
+    _registry_keys = list(PRESETS.keys())
+
+    def init(self,
+             authorization_url=None,
+             token_url=None,
+             auto_refresh_url=None,
+             scopes=None,
+             user_info_endpoint=None,
+             client_id=None,
+             client_secret=None,
+             user_info_mapping=None,
+             validate=None):
+
+        (self.base_url,
+         self.authorization_url,
+         self.token_url,
+         self.auto_refresh_url,
+         self.scopes,
+         self.user_info_endpoint,
+         self.client_id,
+         self.client_secret,
+         self.user_info_mapping,
+         validate) = _resolve_oauth_config(
+            self.name,
+            locals(),
+            self.app.config,
+            'base_url',
+            'authorization_url',
+            'token_url',
+            'auto_refresh_url',
+            'scopes',
+            'user_info_endpoint',
+            'client_id',
+            'client_secret',
+            'user_info_mapping',
+            'validate'
+        )
+        if validate is not None:
+            self.validate = validate
```

Comment:
This implementation implies functions in the server_config.py in the format:

`def OAUTH_{provider name}_VALIDATE(user):`

whereas what we really need is 

`def OAUTH_{provider name}_VALIDATE(provider, user):`

to be able to make use of further oauth_client_get calls.

Could this be replaced by something like `self.validate = lambda x: validate(self, x)`?